EventPhases/explainer.md

```diff
@@ -0,0 +1,501 @@
+> - **Layout**: CSS [layout containment (via `contain: layout;`)](https://developer.mozilla.org/en-US/docs/Web/CSS/contain) delivers subtree isolation in the layout phase, allowing sections of a document to "skip" many global layout operations.<br><br>
+> - **Paint**: CSS paint containment (via [`contain: paint;`](https://developer.mozilla.org/en-US/docs/Web/CSS/contain)) ensures that content within an element cannot require the rasterizer and compositor systems consider changes within the element to impact stacking and painting of other elements (common challenging examples include `box-shadow`).<br><br>
+> - **Composite**: Judicious use of [`will-change`](https://css-tricks.com/almanac/properties/w/will-change/) and strictly defined CSS animations can dramatically impact responsiveness of interactive content, however most compositing operations remain only indirectly influenceable.
+>
+> [`content-visibility`](https://web.dev/articles/content-visibility) also provides another axis of layout and paint-phase work management. By preventing those calculations from occurring for elements that are known _a priori_ to be outside the viewport, `content-visibility` can be combined with other techniques to dramatically reduce work generated by DOM changes.
+
+Large web applications today are generally not built with these new APIs in mind, leading to single-threaded performance cliffs when selector caches reach their limits or are unadvisedly invalidated (e.g., by "CSS-in-JS" runtimes). Combined with the reality that most of today's computers are slow Android phones, this creates consistent challenges for web application responsiveness.
+
+The scale and persistence of these issues suggests missing platform controls to facilitate better page-level outcomes.
+
+### Goals
+
+ - Identify APIs or controls that can provide document authors with page-level control to prevent interleaving of style readbacks and writes.
+ - Minimal API surface area.
+ - Ability to be incrementally adopted.
+ - Integration with the [Reporting API,](https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API) including a [Report-Only mode,](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy-Report-Only) to assist in migrations.
+
+### Non-goals
+
+ - Subtree-scale control for read and write phasing. It will be challenging enough for developers to consider a global regime; fine-grained configuration may harm adoption as much as could enhance it.
+ - New, fully asynchronous style readback APIs (see Considered Alternatives below).
+
+## Proposal
+
+We propose to introduce separate read and write phases when interacting with the DOM. When this mode is enabled, interleaving reads and writes will no longer incidentally trigger global style recalculation and layout, but will either result in "stale" reads or exceptions.
+
+Today's phasing can be described as:
+
+ 1. Browser processing (style, layout, etc.)
+ 1. JavaScript event dispatch
```

Comment:
Maybe add "read/write" somewhere here so it's easier for the reader to connect this phase to the read/write phase below.